openvasd/vtparser.c

```diff
@@ -0,0 +1,414 @@
+    }       // end preferences
+}
+
+/**
+ * @brief Parse a VT element given in json format.
+ *
+ * @param[in] reader JSON reader object pointing to an VT element.
+ *
+ * @return nvti structur containing the VT metadata, NULL otherwise.
+ *              The nvti struct must be free() by the caller
+ */
+nvti_t *
+openvasd_parse_vt (gvm_json_pull_parser_t *parser, gvm_json_pull_event_t *event)
+{
+  nvti_t *nvt = NULL;
+  cJSON *vt_obj = NULL;
+  cJSON *item = NULL;
+  gchar *error_message = NULL;
+
+  gvm_json_pull_parser_next (parser, event);
+
+  // Handle start/end of json array
+  gchar *path = gvm_json_path_to_string (event->path);
+  if (!g_strcmp0 (path, "$") && event->type == GVM_JSON_PULL_EVENT_ARRAY_START)
+    {
+      gvm_json_pull_parser_next (parser, event);
+      g_debug ("%s: Start parsing feed", __func__);
+    }
+  else if (!g_strcmp0 (path, "$")
+           && (event->type == GVM_JSON_PULL_EVENT_ARRAY_END
+               || event->type == GVM_JSON_PULL_EVENT_EOF))
+    {
+      g_debug ("%s: Finish parsing feed", __func__);
+      g_free (path);
+      return NULL;
+    }
+  g_free (path);
+
+  // It is an NVT object
+  if (event->type != GVM_JSON_PULL_EVENT_OBJECT_START)
+    {
+      g_warning ("%s: Error reading VT object", __func__);
+      return NULL;
+    }
+
+  vt_obj = gvm_json_pull_expand_container (parser, &error_message);
+  if (!cJSON_IsObject (vt_obj))
+    {
+      g_free (error_message);
+      cJSON_Delete (vt_obj);
+      return NULL;
+    }
+
+  nvt = nvti_new ();
+
+  if ((item = cJSON_GetObjectItem (vt_obj, "oid")) != NULL
+      && cJSON_IsString (item))
+    nvti_set_oid (nvt, item->valuestring);
+  else
+    {
+      g_warning ("%s: VT missing OID", __func__);
+      cJSON_Delete (vt_obj);
+      g_free (error_message);
+      nvti_free (nvt);
+      return NULL;
+    }
+
+  if ((item = cJSON_GetObjectItem (vt_obj, "name")) != NULL
+      && cJSON_IsString (item))
+    nvti_set_name (nvt, item->valuestring);
+  else
+    {
+      g_warning ("%s: VT missing NAME", __func__);
+      cJSON_Delete (vt_obj);
+      g_free (error_message);
+      nvti_free (nvt);
+      return NULL;
+    }
+
+  if ((item = cJSON_GetObjectItem (vt_obj, "family")) != NULL
+      && cJSON_IsString (item))
+    nvti_set_family (nvt, item->valuestring);
+  else
+    {
+      g_warning ("%s: VT missing FAMILY", __func__);
+      cJSON_Delete (vt_obj);
+      g_free (error_message);
+      nvti_free (nvt);
+      return NULL;
+    }
+
+  if ((item = cJSON_GetObjectItem (vt_obj, "category")) != NULL
+      && cJSON_IsString (item))
+    nvti_set_category (nvt, get_category_from_name (item->valuestring));
+  else
+    {
+      g_warning ("%s: VT missing CATEGORY", __func__);
+      cJSON_Delete (vt_obj);
+      g_free (error_message);
+      nvti_free (nvt);
+      return NULL;
+    }
+
+  cJSON *tag_obj = cJSON_GetObjectItem (vt_obj, "tag");
+  if (tag_obj)
+    add_tags_to_nvt (nvt, tag_obj);
+  if (!nvt)
```

Comment:
Looks like another redundant g_malloc check that was intended to be done straight after the nvti_new.